fml/message_loop_impl.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/message_loop_impl.h"
+#include "lib/ftl/build_config.h"
+
+#if OS_MACOSX
+
+#include "flutter/fml/platform/darwin/message_loop_darwin.h"
+using PlatformMessageLoopImpl = fml::MessageLoopDarwin;
+
+#else
+
+#error This platform does not have a message loop implementation.
+
+#endif
+
+namespace fml {
+
+ftl::RefPtr<MessageLoopImpl> MessageLoopImpl::Create() {
+  return ftl::MakeRefCounted<::PlatformMessageLoopImpl>();
+}
+
+MessageLoopImpl::MessageLoopImpl() : order_(0) {}
+
+MessageLoopImpl::~MessageLoopImpl() = default;
+
+void MessageLoopImpl::PostTask(ftl::Closure task, ftl::TimePoint target_time) {
+  if (task == nullptr) {
+    return;
+  }
+
+  WakeUp(RegisterTaskAndGetNextWake(task, target_time));
+}
+
+void MessageLoopImpl::RunExpiredTasksNow() {
+  WakeUp(RunExpiredTasksAndGetNextWake());
+}
+
+void MessageLoopImpl::SetTaskObserver(MessageLoop::TaskObserver observer) {
+  ftl::MutexLocker lock(&task_observer_mutex_);
+  task_observer_ = observer;
+}
+
+ftl::TimePoint MessageLoopImpl::RegisterTaskAndGetNextWake(
+    ftl::Closure task,
+    ftl::TimePoint target_time) {
+  FTL_DCHECK(task != nullptr);
+  ftl::MutexLocker lock(&delayed_tasks_mutex_);
+  delayed_tasks_.push({++order_, std::move(task), target_time});
+  return delayed_tasks_.top().target_time;
+}
+
+ftl::TimePoint MessageLoopImpl::RunExpiredTasksAndGetNextWake() {
+  std::vector<ftl::Closure> invocations;
+
+  {
+    ftl::MutexLocker lock(&delayed_tasks_mutex_);
+
+    if (delayed_tasks_.empty()) {
+      return ftl::TimePoint::Max();
+    }
+
+    auto now = ftl::TimePoint::Now();
+    while (!delayed_tasks_.empty()) {
```

Comment:
Turns out there was one spot where delayed tasks were used. It was in the fallback VSync waiter on Linux. Admittedly, that is not used on all platforms, but it seemed icky that I would have to wire up delayed tasks on one platform and not others. Just implemented it for all platforms instead.